APIT-3126: Add disable_wait_for_ready attribute to disable readiness checks for confluent_role_binding resource (#657)
confluentinc/terraform-provider-confluent: docs/resources/confluent_role_binding.md

- `role_name` - (Required String) A name of the role to bind to the principal. See [Confluent Cloud RBAC Roles](https://docs.confluent.io/cloud/current/access-management/access-control/cloud-rbac.html#ccloud-rbac-roles) for a full list of supported role names.
- `crn_pattern` - (Required String) A [Confluent Resource Name (CRN)](https://docs.confluent.io/cloud/current/api.html#section/Identifiers-and-URLs/Confluent-Resource-Names-(CRNs)) that specifies the scope and resource patterns necessary for the role to bind.
- `disable_wait_for_ready` - (Optional Boolean) An optional flag to disable wait-for-readiness on create. Its primary use case is to ensure that the role bindings are properly synchronized across the system before exiting the Terraform creation step. Must be unset when importing. Defaults to `false`.

!> **Warning:** When `disable_wait_for_ready = true` is used, Terraform skips waiting for role bindings to fully propagate. This can lead to a situation where Terraform attempts to create resources before the service account has the necessary permissions—resulting in HTTP 403 Forbidden errors.

> For example, if you're creating a new service account, a new Kafka API Key, a new CloudClusterAdmin role binding, and a Kafka topic in a single run (see this [code snippet](https://github.com/confluentinc/terraform-provider-confluent/blob/master/examples/configurations/standard-kafka-rbac/main.tf#L46-L100)), the topic creation may fail if the role binding hasn’t taken effect yet. Without that role, the service account won’t have permission to create the topic.

Reviewer: nit:

Maybe `CloudClusterAdmin` since this is a RBAC predefined term.

Author: That's a great idea! I applied the suggestion, thanks!

This setting is best suited for scenarios where you're provisioning a large number of role bindings without dependent resources, as it significantly speeds up the apply process.

> -> **Note:** If you encounter HTTP 403 Forbidden errors when creating role bindings, feel free to rerun `terraform apply` after a few minutes, once the role bindings have had time to propagate.

Reviewer: nit:

feel free to rerun -> you can rerun

It sounds more formal with a more neutral tone for technical documentation.

Author: That's a great idea! I applied the suggestion, thanks!

> -> **Note:** Feel free to use `time_sleep` [resource](https://registry.terraform.io/providers/hashicorp/time/latest/docs/resources/sleep) to configure a custom waiting period, see [this example](#example-of-using-time_sleep) for more details.

Reviewer: ```suggestion
-> **Note:** You can also use `time_sleep` [resource](https://registry.terraform.io/providers/hashicorp/time/latest/docs/resources/sleep) to configure a custom waiting period, see [this example](#example-of-using-time_sleep) for more details.
```

Reviewer: Maybe we can explicitly call out that this is a HashiCorp resource (which can be utilized to achieve our goal)?

  * [`dedicated-transit-gateway-attachment-aws-kafka-acls`](https://github.com/confluentinc/terraform-provider-confluent/tree/master/examples/configurations/dedicated-transit-gateway-attachment-aws-kafka-acls): _Dedicated_ Kafka cluster on AWS that is accessible via Transit Gateway Endpoint with authorization using ACLs
  * [`dedicated-transit-gateway-attachment-aws-kafka-rbac`](https://github.com/confluentinc/terraform-provider-confluent/tree/master/examples/configurations/dedicated-transit-gateway-attachment-aws-kafka-rbac): _Dedicated_ Kafka cluster on AWS that is accessible via Transit Gateway Endpoint with authorization using RBAC
  * [`enterprise-privatelinkattachment-aws-kafka-acls`](https://github.com/confluentinc/terraform-provider-confluent/tree/master/examples/configurations/enterprise-privatelinkattachment-aws-kafka-acls): _Enterprise_ Kafka cluster on AWS that is accessible via PrivateLink connections with authorization using ACLs

## Example of using time_sleep

> This configuration introduces a 10-second custom delay after the creation of a role binding, before creating a Kafka topic.

Reviewer: nit:

```suggestion
This configuration introduces a 10-second custom delay after the creation of a role binding, allowing time for the RBAC role binding to fully propagate before creating a Kafka topic.
```

Author: > This configuration introduces a 10-second custom delay after the creation of a role binding, allowing time for the RBAC role binding to fully propagate before creating a Kafka topic.

10 seconds won't be enough to fully propagate actually 😁 Based on our test, it takes ~10 seconds for a role binding to propagate for a control place resource (Kafka cluster, environment), but it can take way longer (1+ minute) for data plane resources like Kafka topic.

Author: Update: I have updated the example to include a 360-second custom delay, which will be applicable if our default waiting mechanism fails, and a user would like to extend the waiting time.

```terraform
resource "confluent_role_binding" "app-manager-kafka-cluster-admin-skip-sync" {
  principal   = "User:${confluent_service_account.app-manager.id}"
  role_name   = "CloudClusterAdmin"
  crn_pattern = confluent_kafka_cluster.standard.rbac_crn
  disable_wait_for_ready = true
}

> resource "time_sleep" "wait_10_seconds_after_role_binding" {

Reviewer: I suppose this is the HashiCorp native resource to block the remaining resources to be created as an alternative approach?

Author: Yes, this is a very custom use case that we don't expect customers to use often, but it's just an example of how it could be configured. So, we won't have to introduce another resource to accept a time delay, and a boolean attribute is sufficient here.

## Example of using time_sleep

> This configuration introduces a 360-second custom delay after the creation of a role binding, before creating a Kafka topic.

Reviewer: (Can approve this PR anytime)

No concern on the documentation or examples, if I were a user, I might be wondering why we choose the 360s magic number, and if there is hidden logic.

Let's continue the discussion here if not urgent:
https://confluent.slack.com/archives/C075H73ATEX/p1745449146222049?thread_ts=1745341487.349369&cid=C075H73ATEX

Author: That's a great question! 360 is 5 (their team's SLO) + 1 extra minute as a buffer, but I didn't want to expose internal details 😀
